go-controller/pkg/ovn/egressqos.go

```diff
@@ -0,0 +1,909 @@
+
+func (oc *Controller) processNextEgressQoSWorkItem(wg *sync.WaitGroup) bool {
+	wg.Add(1)
+	defer wg.Done()
+
+	key, quit := oc.egressQoSQueue.Get()
+	if quit {
+		return false
+	}
+
+	defer oc.egressQoSQueue.Done(key)
+
+	err := oc.syncEgressQoS(key.(string))
+	if err == nil {
+		oc.egressQoSQueue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	if oc.egressQoSQueue.NumRequeues(key) < maxEgressQoSRetries {
+		oc.egressQoSQueue.AddRateLimited(key)
+		return true
+	}
+
+	oc.egressQoSQueue.Forget(key)
+	return true
+}
+
+// This takes care of syncing stale data which we might have in OVN if
+// there's no ovnkube-master running for a while.
+// It deletes all QoSes and Address Sets from OVN that belong to EgressQoSes.
+func (oc *Controller) repairEgressQoSes() error {
+	startTime := time.Now()
+	klog.V(4).Infof("Starting repairing loop for egressqos")
+	defer func() {
+		klog.V(4).Infof("Finished repairing loop for egressqos: %v", time.Since(startTime))
+	}()
+
+	p := func(q *nbdb.QoS) bool {
+		_, ok := q.ExternalIDs["EgressQoS"]
+		return ok
+	}
+	existingQoSes, err := libovsdbops.FindQoSesWithPredicate(oc.nbClient, p)
+	if err != nil {
+		return err
+	}
+
+	if len(existingQoSes) > 0 {
+		allOps := []ovsdb.Operation{}
+
+		ops, err := libovsdbops.DeleteQoSesOps(oc.nbClient, nil, existingQoSes...)
+		if err != nil {
+			return err
+		}
+		allOps = append(allOps, ops...)
+
+		logicalSwitches, err := oc.egressQoSSwitches()
+		if err != nil {
+			return err
+		}
+
+		for _, sw := range logicalSwitches {
+			ops, err := libovsdbops.RemoveQoSesFromLogicalSwitchOps(oc.nbClient, nil, sw, existingQoSes...)
+			if err != nil {
+				return err
+			}
+			allOps = append(allOps, ops...)
+		}
+
+		if _, err := libovsdbops.TransactAndCheck(oc.nbClient, allOps); err != nil {
+			return fmt.Errorf("unable to remove stale qoses, err: %v", err)
+		}
+	}
+
+	asPredicate := func(as *nbdb.AddressSet) bool {
+		return strings.Contains(as.ExternalIDs["name"], types.EgressQoSRulePrefix)
+	}
+	if err := libovsdbops.DeleteAddressSetsWithPredicate(oc.nbClient, asPredicate); err != nil {
+		return fmt.Errorf("failed to remove stale egress qos address sets, err: %v", err)
+	}
+
+	return nil
+}
+
+func (oc *Controller) syncEgressQoS(key string) error {
+	startTime := time.Now()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Processing sync for EgressQoS %s/%s", namespace, name)
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing EgressQoS %s on namespace %s : %v", name, namespace, time.Since(startTime))
+	}()
+
+	eq, err := oc.egressQoSLister.EgressQoSes(namespace).Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if name != defaultEgressQoSName {
+		klog.Errorf("EgressQoS name %s is invalid, must be %s", name, defaultEgressQoSName)
+		return nil // Return nil to avoid requeues
+	}
+
+	err = oc.cleanEgressQoSNS(namespace)
```

Comment:
so everytime we sync we delete and add? This could result in temporary loss of marking packets on pods. In the future it would be better to do the clean and add as a single ovsdb operation?

Reply:
yeah I think it's possible, but we'd need to cover the situation where the updated resource fails - clean only / keep the old state until we update, etc.

Reply:
ok for now its fine, but can you add a // TODO here explaining we should reconcile better to avoid windows dscp marking isn't working?

Reply:
done